File: 0x13-javascript_objects_scopes_closures/4-rectangle.js
#!/usr/bin/node

class Rectangle {
  constructor (w, h) {
    if (w > 0 && h > 0) {
      this.width = w;
      this.height = h;
    }
  }

  print (w, h) {
    let rectangle = '';
    for (let num1 = 0; num1 < this.height; num1++) {
      for (let num2 = 0; num2 < this.width; num2++) {
        rectangle += 'X';
      }
      if (num1 === this.height - 1) {
        continue;
      } else {
        rectangle += '\n';
      }
    }
    console.log(rectangle);
  }

  double (w, h) {
    this.width = this.width * 2;
    this.height = this.height * 2;
  }

  rotate (w, h) {
    const proxy = this.width;
    this.width = this.height;
    this.height = proxy;
  }
}

module.exports = Rectangle;

File: 0x08-python-more_classes/4-rectangle.py
#!/usr/bin/python3
"""
A class Rectangle that defines a rectangle.
"""


class Rectangle():
    """class that defines a rectangle based on 2-rectangle.py

        Attributes:
            width (int): width of rectangle
            height (int): height of rectangle
    """
    __width = None
    __height = None

    def __init__(self, width=0, height=0):
        """initialise the Rectangle class
        Args:
            width (int, optional): width of rectangle.
            height (int, optional): height of rectangle .
        """
        self.__width = width
        self.__height = height

    @property
    def width(self):
        """
        @property -gets width of rectangle
        @width.setter - validates and sets width

        Returns:
            width(int)
        """
        return self.__width

    @width.setter
    def width(self, value):
        if not isinstance(value, int):
            raise TypeError("width must be an integer")
        if value < 0:
            raise ValueError("width must be >= 0")

        self.__width = value

    @property
    def height(self):
        """
        @property -gets height of rectangle
        @width.setter - validates and sets height

        Returns:
            height(int)
        """
        return self.__height

    @height.setter
    def height(self, value):
        if not isinstance(value, int):
            raise TypeError("height must be an integer")
        if value < 0:
            raise ValueError("height must be >= 0")

        self.__height = value

    def area(self):
        """calculate area of rectangle

        Returns:
            area (int): area of rectangle
        """
        return self.__height * self.__width

    def perimeter(self):
        """calculate perimeter of rectangle

        Returns:
            perimeter(int): perimeter of rectangles
        """
        perimeter = 0
        if self.__width == 0 or self.__height == 0:
            return perimeter
        else:
            return (self.__height + self.__width) * 2

    def __str__(self):
        """returns a string of the rectangle
        with the '#' character
        """
        rect = ""
        if self.__width == 0 or self.__height == 0:
            return rect

        for num in range(0, self.__height):
            for num2 in range(0, self.__width):
                rect += "#"
            if num == self.__height - 1:
                continue
            else:
                rect += "\n"

        return rect

    def __repr__(self):
        """ return a string representation of the rectangle
        """
        rect_str = "Rectangle(" + str(self.__width) + ", " + \
            str(self.__height) + ")"

        return rect_str
